File: Java/src/Color/ColorRGBA.java
package Color;

public class ColorRGBA extends ColorRGB {
    private short alpha;

    public ColorRGBA() {
        super();
        alpha = 0;
    }

    public ColorRGBA(short a) {
        super();
        alpha = (short) (Math.abs(a) % 256);
    }

    public ColorRGBA(short r, short a) {
        super(r);
        alpha = (short) (Math.abs(a) % 256);
    }

    public ColorRGBA(short r, short g, short a) {
        super(r, g);
        alpha = (short) (Math.abs(a) % 256);
    }

    public ColorRGBA(short r, short g, short b, short a) {
        super(r, g, b);
        alpha = (short) (Math.abs(a) % 256);
    }

    public ColorRGBA(ColorRGBA crgba) {
        super(crgba);
        alpha = crgba.alpha;
    }

    public short getAlpha() {
        return alpha;
    }

    public ColorRGBA sum(ColorRGB crgba) {
        ColorRGB tmp = super.sum(crgba);
        if (getClass() == crgba.getClass()) {
            ColorRGBA c = (ColorRGBA) crgba;
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), (short) (alpha + c.alpha));
        } else {
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), alpha);
        }
    }

    public ColorRGBA sub(ColorRGB crgba) {
        ColorRGB tmp = super.sub(crgba);
        if (getClass() == crgba.getClass()) {
            ColorRGBA c = (ColorRGBA) crgba;
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), (short) (alpha - c.alpha));
        } else {
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), alpha);
        }
    }

    public ColorRGBA mul(ColorRGB crgba) {
        ColorRGB tmp = super.mul(crgba);
        if (getClass() == crgba.getClass()) {
            ColorRGBA c = (ColorRGBA) crgba;
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), (short) (alpha * c.alpha));
        } else {
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), alpha);
        }
    }


    public ColorRGBA div(ColorRGB crgba) {
        ColorRGB tmp = super.div(crgba);
        if (getClass() == crgba.getClass()) {
            ColorRGBA c = (ColorRGBA) crgba;
            if (alpha == 0)
                return new ColorRGBA(this);
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), (short) (alpha / c.alpha));
        } else {
            return new ColorRGBA(tmp.getRed(), tmp.getGreen(), tmp.getBlue(), alpha);
        }
    }

    public boolean equals(Object obj) {
        if (super.equals(obj)) {
            ColorRGBA crgba = (ColorRGBA) obj;
            return alpha == crgba.alpha;
        }
        return false;
    }

    public String toString() {
        return "ColorRGBA(" + getRed() + "," + getGreen() + "," + getBlue() + "," + alpha + ")";
    }
}

File: cpp/KalkCpp/Color/colorrgba.cpp
#include "colorrgba.h"
#include <typeinfo>



ColorRGBA::ColorRGBA(unsigned short r, unsigned short g, unsigned short b, unsigned short a) : ColorRGB(r,g,b), alpha(a)
{
    alpha = abs(alpha) % 256;
}


bool ColorRGBA::operator==(const ColorRGB& crgb) const
{
    if(ColorRGB::operator ==(crgb))
    {
        const ColorRGBA& crgba = static_cast<const ColorRGBA&>(crgb);
        if(alpha == crgba.alpha)
        {
            return true;
        }
    }
    return false;
}







ColorRGBA* ColorRGBA::clone() const
{
    return new ColorRGBA(*this);
}


ColorRGBA* ColorRGBA::operator-(const ColorRGB& crgba) const
{
    ColorRGB* temp(ColorRGB::operator -(crgba));
    if(typeid(*this) == typeid(crgba))
    {
        const ColorRGBA& t = static_cast<const ColorRGBA&>(crgba);
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha - t.alpha);
    }
    else
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha);
}



ColorRGBA* ColorRGBA::operator+(const ColorRGB& crgba) const
{
    ColorRGB* temp(ColorRGB::operator +(crgba));
    if(typeid(*this) == typeid(crgba))
    {
        const ColorRGBA& t = static_cast<const ColorRGBA&>(crgba);
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha + t.alpha);
    }
    else
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha);
}



ColorRGBA* ColorRGBA::operator*(const ColorRGB& crgba) const
{
    ColorRGB* temp(ColorRGB::operator *(crgba));
    if(typeid(*this) == typeid(crgba))
    {
        const ColorRGBA& t = static_cast<const ColorRGBA&>(crgba);
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha * t.alpha);
    }
    else
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha);
}



ColorRGBA* ColorRGBA::operator/(const ColorRGB& crgba) const
{
    ColorRGB* temp(ColorRGB::operator /(crgba));
    if(typeid(*this) == typeid(crgba))
    {
        const ColorRGBA& t = static_cast<const ColorRGBA&>(crgba);
        if(t.alpha == 0)
            return this->clone();
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha / t.alpha);
    }
    else
        return new ColorRGBA(temp->getRed(), temp->getGreen(), temp->getBlue(), alpha);
}



void ColorRGBA::printColor() const
{
    std::cout << "ColorRGBA(" + std::to_string(getRed()) + "," + std::to_string(getGreen()) + "," + std::to_string(getBlue()) + "," + std::to_string(alpha) + ")";
}



std::ostream& operator<<(std::ostream& os, const ColorRGBA& crgba)
{
   return os<<"("<<crgba.getRed()<<","<<crgba.getGreen()<<","<<crgba.getBlue()<<"," << crgba.alpha << ")";
}
